server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java

```diff
@@ -0,0 +1,449 @@
+                    () -> new ParameterizedMessage(
+                        "failed to start decommission action for attribute [{}]",
+                        decommissionAttribute.toString()
+                    ),
+                    e
+                );
+                listener.onFailure(e);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                DecommissionAttributeMetadata decommissionAttributeMetadata = newState.metadata()
+                    .custom(DecommissionAttributeMetadata.TYPE);
+                assert decommissionAttribute.equals(decommissionAttributeMetadata.decommissionAttribute());
+                if (decommissionAttributeMetadata.status().equals(DecommissionStatus.SUCCESSFUL)) {
+                    logger.info("status is already marked SUCCESSFUL, no need to proceed for further processing");
+                    listener.onResponse(new ClusterStateUpdateResponse(true));
+                } else {
+                    decommissionClusterManagerNodes(decommissionAttributeMetadata.decommissionAttribute(), listener);
+                }
+            }
+        });
+    }
+
+    private synchronized void decommissionClusterManagerNodes(
+        final DecommissionAttribute decommissionAttribute,
+        ActionListener<ClusterStateUpdateResponse> listener
+    ) {
+        ClusterState state = clusterService.getClusterApplierService().state();
+        Set<DiscoveryNode> clusterManagerNodesToBeDecommissioned = filterNodesWithDecommissionAttribute(state, decommissionAttribute, true);
+        // This check doesn't seem to be needed as exclusion automatically shrinks the config before sending the response.
+        // We can guarantee that because of exclusion there wouldn't be a quorum loss and if the service gets a successful response,
+        // we are certain that the config is updated and nodes are ready to be kicked out.
+        // Please add comment if you feel there could be a edge case here.
+        // try {
+        // // this is a sanity check that the cluster will not go into a quorum loss state because of exclusion
+        // ensureNoQuorumLossDueToDecommissioning(
+        // decommissionAttribute,
+        // clusterManagerNodesToBeDecommissioned,
+        // state.getLastCommittedConfiguration()
+        // );
+        // } catch (DecommissioningFailedException dfe) {
+        // listener.onFailure(dfe);
+        // decommissionController.updateMetadataWithDecommissionStatus(DecommissionStatus.FAILED, statusUpdateListener());
+        // return;
+        // }
+
+        ActionListener<Void> exclusionListener = new ActionListener<Void>() {
+            @Override
+            public void onResponse(Void unused) {
+                if (transportService.getLocalNode().isClusterManagerNode()
+                    && !nodeHasDecommissionedAttribute(transportService.getLocalNode(), decommissionAttribute)) {
```

Comment:
Updated the logic to below - 
```
if (clusterService.getClusterApplierService().state().nodes().isLocalNodeElectedClusterManager()) {
                    logger.info("will attempt to fail decommissioned nodes as local node is eligible to process the request");
                    if (nodeHasDecommissionedAttribute(transportService.getLocalNode(), decommissionAttribute)) {
                        // this is an unexpected state, as after exclusion of nodes having decommission attribute,
                        // this local node shouldn't have had the decommission attribute. Will send the failure response to the user
                        listener.onFailure(new IllegalStateException(
                            "unexpected state encountered [local node is to-be-decommissioned leader] while executing decommission request"
                            )
                        );
                    } else {
                        // we are good here to send the response now as the request is processed by an eligible active leader
                        // and to-be-decommissioned cluster manager is no more part of Voting Configuration
                        listener.onResponse(new ClusterStateUpdateResponse(true));
                        failDecommissionedNodes(clusterService.getClusterApplierService().state());
                    }
                } else {
                    // explicitly calling listener.onFailure with NotClusterManagerException as the local node is not the cluster manager
                    // this will ensures that request is retried until cluster manager times out
                    logger.info(
                        "local node is not eligible to process the request, "
                            + "throwing NotClusterManagerException to attempt a retry on an eligible node"
                    );
                    listener.onFailure(
                        new NotClusterManagerException(
                            "node ["
                                + transportService.getLocalNode().toString()
                                + "] not eligible to execute decommission request. Will retry until timeout."
                        )
                    );
                }
```